---
layout: default
title: {{ site.name }}
---

<p>GBRS is a suite of tools for reconstructing genomes using RNA-Seq data from multiparent population and quantifying allele specific expression. Although we tested it with mouse models only, GBRS should work for any multiparent populations. If you are interested in using it for other multiparent population model, please <a href="mailto:kb.choi@jax.org?Subject=[GBRS]%20Question" target="_top">contact</a> Kwangbom “KB” Choi at The Jackson Laboratory. For the Diversity Outbred (<a href="https://www.jax.org/strain/009376">DO</a>), Collaborative Cross (CC) mice, or F1 hybrids of CC’s (CCRIX), required data files are available at <a href="ftp://churchill-lab.jax.org/software/GBRS/">ftp://churchill-lab.jax.org/software/GBRS/</a>.</p>

<ul>
  <li>Free software: GPLv3 license</li>
  <li>Full documentation: <a href="https://gbrs.readthedocs.io/en/latest/readme.html">https://gbrs.readthedocs.io/en/latest/readme.html</a></li>
</ul>

<h2>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Installation</h2>

<p><strong>Note:</strong> Although GBRS is available at <a href="https://pypi.python.org/pypi/gbrs/">PyPI</a> for <code>pip install</code> or <code>easy_install</code>, we highly recommend using <a href="https://www.continuum.io/downloads">Anaconda distribution</a> of python to install all the dependencies without issues. GBRS is also available on <a href="https://anaconda.org/kbchoi/gbrs">Anaconda Cloud</a>, so add the following channels if you have not already:</p>

<pre>
  <code>$ conda config --add channels r
  $ conda config --add channels bioconda</code>
</pre>

<p>To avoid conflicts among dependencies, we also highly recommend using conda virtual environment:</p>

<pre>
  <code>$ conda create -n gbrs python=2
  $ source activate gbrs</code>
</pre>

<p>Once GBRS virtual environment is created and activated, your shell prompt will show ‘(gbrs)’ at the beginning to specify what virtual environment you are currently in. Now please type the following and install GBRS:</p>

<pre><code>(gbrs) $ conda install -c kbchoi gbrs
</code></pre>

<p>That's all! Now gbrs is available. Once you are done using GBRS, you can go out from GBRS virtual environment anytime:</p>

<pre><code>(gbrs) $ source deactivate
</code></pre>

<h2><a id="usage" class="anchor" href="#usage" aria-hidden="true"><span class="octicon octicon-link"></span></a>Usage</h2>

<p><strong>Note:</strong> We will assume you installed GBRS in its own conda virtual environment. First of all, you have to “activate” the virtual environment by doing the following:</p>

<pre><code>$ source activate gbrs</code></pre>

<p>The first step is to align our RNA-Seq reads against pooled transcriptome of all founder strains:</p>

<pre><code>$ bowtie -q -a --best --strata --sam -v 3 ${GBRS_DIR}/bowtie.transcriptome ${FASTQ} \
    | samtools view -bS - > ${BAM_FILE}</code>
</pre>

<p>Before quantifying multiway allele specificity, bam file should be converted into emase format:</p>

<pre><code>$ gbrs bam2emase -i ${BAM_FILE} \
                 -m ${GBRS_DATA}/ref.transcripts.info \
                 -s ${COMMA_SEPARATED_LIST_OF_HAPLOTYPE_CODES} \
                 -o ${EMASE_FILE}</code>
</pre>

<p>We can compress EMASE format alignment incidence matrix:</p>

<pre><code>$ gbrs compress -i ${EMASE_FILE} \
                -o ${COMPRESSED_EMASE_FILE}</code>
</pre>

<p>If storage space is tight, you may want to delete ${BAM_FILE} or ${EMASE_FILE} at this point since ${COMPRESSED_EMASE_FILE} has all the information the following steps would need. If you want to merge emase format files in order to, for example, pool technical replicates, you run ‘compress’ once more listing files you want to merge with commas:</p>

<pre><code>$ gbrs compress -i ${COMPRESSED_EMASE_FILE1},${COMPRESSED_EMASE_FILE2},... \
                -o ${MERGED_COMPRESSED_EMASE_FILE}</code>
</pre>

<p>and use ${MERGED_COMPRESSED_EMASE_FILE} in the following steps. Now we are ready to quantify multiway allele specificity:</p>

<pre><code>$ gbrs quantify -i ${COMPRESSED_EMASE_FILE} \
                -g ${GBRS_DATA}/ref.gene2transcripts.tsv \
                -L ${GBRS_DATA}/gbrs.hybridized.targets.info \
                -M 4 --report-alignment-counts</code>
</pre>

<p>Then, we reconstruct the genome based upon gene-level TPM quantities (assuming the sample is a female from the 20th generation Diversity Outbred mice population)</p>

<pre><code>$ gbrs reconstruct -e gbrs.quantified.multiway.genes.tpm \
                   -t ${GBRS_DATA}/tranprob.DO.G20.F.npz \
                   -x ${GBRS_DATA}/avecs.npz \
                   -g ${GBRS_DATA}/ref.gene_pos.ordered.npz</code>
</pre>

<p>We can now quantify allele-specific expressions on diploid transcriptome:</p>

<pre><code>$ gbrs quantify -i ${COMPRESSED_EMASE_FILE} \
                -G gbrs.reconstructed.genotypes.tsv \
                -g ${GBRS_DATA}/ref.gene2transcripts.tsv \
                -L ${GBRS_DATA}/gbrs.hybridized.targets.info \
                -M 4 --report-alignment-counts</code>
</pre>

<p>Genotype probabilities are on a grid of genes. For eQTL mapping or plotting genome reconstruction, we may want to interpolate probability on a decently-spaced grid of the reference genome.:</p>

<pre><code>$ gbrs interpolate -i gbrs.reconstructed.genoprobs.npz \
                   -g ${GBRS_DATA}/ref.genome_grid.64k.txt \
                   -p ${GBRS_DATA}/ref.gene_pos.ordered.npz \
                   -o gbrs.interpolated.genoprobs.npz</code>
</pre>

<p>To plot a reconstructed genome:</p>

<pre><code>$ gbrs plot -i gbrs.interpolated.genoprobs.npz \
            -o gbrs.plotted.genome.pdf \
            -n ${SAMPLE_ID}</code>
</pre>


<div id="home">
  <h2>Announcements</h2>
  <ul class="posts">
    {% for post in site.posts %}
      <li><span>{{ post.date | date_to_string }}</span> &raquo; <a href="{{ site.github.url }}{{ post.url }}">{{ post.title }}</a></li>
    {% endfor %}
  </ul>
</div>