====
GBRS
====

.. image:: https://img.shields.io/pypi/v/gbrs.svg
        :target: https://pypi.python.org/pypi/gbrs

.. image:: https://anaconda.org/kbchoi/gbrs/badges/version.svg
    :target: https://anaconda.org/kbchoi/gbrs

.. image:: https://travis-ci.org/churchill-lab/gbrs.svg?branch=master
    :target: https://travis-ci.org/churchill-lab/gbrs

.. image:: https://readthedocs.org/projects/gbrs/badge/?version=latest
        :target: https://readthedocs.org/projects/gbrs/?badge=latest
        :alt: Documentation Status

.. image:: https://zenodo.org/badge/54070005.svg
        :target: https://zenodo.org/badge/latestdoi/54070005

Overview
--------

**GBRS** is a suite of tools for reconstructing genomes using RNA-Seq data from multiparent population and quantifying allele specific expression. Although we tested it with mouse models only, GBRS should work for any multiparent populations. If you are interested in using it for other multiparent population model, please contact Kwangbom "KB" Choi <kb.choi@jax.org> at The Jackson Laboratory. For the Diversity Outbred (DO_), Collaborative Cross (CC_) mice, or F1 hybrids of CC's (CCRIX), required data files are available at ftp://churchill-lab.jax.org/software/GBRS/.

.. _DO: https://www.jax.org/strain/009376
.. _CC: https://www.jax.org/mouse-search/?straingroup=Collaborative%20Cross

* Free software: MIT license
* Documentation: https://gbrs.readthedocs.org.


References
----------

Manuscript in preparation (Expected in 2020)
